zebra: Recheck route availability when kernel removed route.

imzyxwvu opened the pull request:
When an interface goes down, kernel notifies zebra that routes with
the interface as nexthop are being deleted. EHOSTUNREACH would occur
if zebra tries to reinstall them. Lots of such log will appear:

```netlink-dp (NS 0) error: No route to host, type=RTM_NEWROUTE(24), seq=69980, pid=2259751423```

Signed-off-by: zyxwvu Shi <i@shiyc.cn>

donaldsharp commented:
This PR makes no sense to me.  FRR does not receive notification about route deletion when a interface goes down:

```
eva# conf
eva(config)# ip route 4.5.6.8/32 10.4.1.2
eva(config)# 4.5.6.8 nhid 387 via 10.4.1.2 dev dummy302 proto static metric 20 

eva(config)# exit
eva# exit
sharpd@eva /t/topotests> sudo ip link set dummy302 down ; sudo ip link set dummy302 up
Deleted fe80::/64 dev dummy302 proto kernel metric 256 pref medium
Deleted anycast fe80:: dev dummy302 table local proto kernel metric 0 pref medium
Deleted local fe80::a84a:edff:fe95:9f18 dev dummy302 table local proto kernel metric 0 pref medium
Deleted multicast ff00::/8 dev dummy302 table local proto kernel metric 256 pref medium
10.4.1.0/24 dev dummy302 proto kernel scope link src 10.4.1.1 
broadcast 10.4.1.0 dev dummy302 table local proto kernel scope link src 10.4.1.1 
broadcast 10.4.1.255 dev dummy302 table local proto kernel scope link src 10.4.1.1 
multicast ff00::/8 dev dummy302 table local proto kernel metric 256 pref medium
fe80::/64 dev dummy302 proto kernel metric 256 pref medium
local fe80::a84a:edff:fe95:9f18 dev dummy302 table local proto kernel metric 0 pref medium
anycast fe80:: dev dummy302 table local proto kernel metric 0 pref medium
sharpd@eva /t/topotests> 
```

The order of events is that FRR receives a link down event and that is the assumed clearing of the route.  What kernel are you using?   And do you have private patches?

imzyxwvu commented:
I have done some additional test and found that this is only related to IPv6 routes.

Reproduction on kernel 5.8:
```
$ sudo ip address add fd01::1/32 dev dummy0
$ sudo ip link set dev dummy0 up
$ sudo ip route add fd02::/16 via fd01::2 dev dummy0
$ sudo ip -6 route show dev dummy0
fd01::/32 proto kernel metric 256 pref medium
fd02::/16 via fd01::2 metric 1024 pref medium
fe80::/64 proto kernel metric 256 pref medium
```

Then I did this in another terminal:
```
# ip link set dev dummy0 down && ip link set dev dummy0 up
```

The netlink events:
```
$ sudo ip monitor link route dev dummy0
3: dummy0: <BROADCAST,NOARP> mtu 1500 qdisc noqueue state DOWN group default 
    link/ether ea:2c:56:aa:f1:9f brd ff:ff:ff:ff:ff:ff
Deleted fd01::/32 proto kernel metric 256 pref medium
Deleted fd02::/16 via fd01::2 metric 1024 pref medium
Deleted fe80::/64 proto kernel metric 256 pref medium
Deleted local fd01::1 table local proto kernel metric 0 pref medium
Deleted local fe80::e82c:56ff:feaa:f19f table local proto kernel metric 0 pref medium
Deleted ff00::/8 table local metric 256 pref medium
3: dummy0: <BROADCAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN group default 
    link/ether ea:2c:56:aa:f1:9f brd ff:ff:ff:ff:ff:ff
192.168.5.0/24 proto kernel scope link src 192.168.5.0 
broadcast 192.168.5.0 table local proto kernel scope link src 192.168.5.0 
broadcast 192.168.5.255 table local proto kernel scope link src 192.168.5.0 
ff00::/8 table local metric 256 pref medium
fe80::/64 proto kernel metric 256 pref medium
local fe80::e82c:56ff:feaa:f19f table local proto kernel metric 0 pref medium
169.254.0.0/16 scope link metric 1000 
```

You can see `Deleted fd02::/16 via fd01::2 metric 1024 pref medium`. This is the route I manually added with `ip route add`.

The behavior is originally found on a device running kernel 4.9.

donaldsharp commented:
Please set `net.ipv6.route.skip_notify_on_dev_down=1` in your sysctl settings.  

In general the code in this section was written to handle the case where some nice operator deleted the FRR route in the linux kernel.  So we could turn around and just re-install it.  It was not meant to handle route deletion due to an interface being shut down.  So where is interface deletion supposed to be handled?  The installing daemon should handle it with the removal or modification of the route in question.

Setting the above sysctl will make v4 and v6 route notification from the kernel behave the same.

If you think that this patch must be pursued the appropriate way to handle it would be to actually re-enqueue the route in the meta-queue instead of calling the nexthop_active where you did.  It breaks some other assumptions about that code that I am not terribly interested in revisiting.

imzyxwvu commented:
Closing this thread since I don't have good idea to resolve the problem.

I could not set `net.ipv6.route.skip_notify_on_dev_down=1`, because I am running an older version of kernel.
